src/hotspot/share/opto/rangeinference.cpp

```diff
@@ -0,0 +1,929 @@
+// significant, and a bit comes after (being lower than) another if it is less
+// significant. For a value n with w bits, we denote n[0] the first (highest)
+// bit of n, n[1] the second bit, ..., n[w - 1] the last (lowest) bit of n.
+template <class U>
+static U adjust_lo(U lo, const KnownBits<U>& bits) {
+  // Violation of lo with respects to bits
+  // E.g: lo    = 1100
+  //      zeros = 0100
+  //      ones  = 1001
+  // zero_violation = 0100, i.e the second bit should be zero, but it is 1 in
+  // lo. Similarly, one_violation = 0001, i.e the last bit should be one, but
+  // it is 0 in lo. These make lo not satisfy the bit constraints, which
+  // results in us having to find the smallest value that satisfies bits.
+  U zero_violation = lo & bits._zeros;
+  U one_violation = ~lo & bits._ones;
+  if (zero_violation == one_violation) {
+    // This means lo does not violate bits, it is the result
+    assert(zero_violation == U(0), "");
+    return lo;
+  }
+
+  /*
+  1. Intuition:
+  Call r the lowest value not smaller than lo that satisfies bits, consider the
+  first bit in r that is different from the corresponding bit in lo:
+    - Since r is larger than lo the bit must be 0 in lo and 1 in r
+    - Since r must satisify bits the bit must be 0 in zeros
+    - Since r should be the smallest value, this bit should be the lowest one
+      possible
+
+  E.g:      1 2 3 4 5 6
+       lo = 1 0 0 1 1 0
+        x = 1 0 1 0 1 0
+        y = 0 1 1 1 1 1
+  x would be larger than lo since the first different bit is the 3rd one,
+  while y is smaller than lo because the first different bit is the 1st bit.
+  Next, consider:
+       x1 = 1 0 1 0 1 0
+       x2 = 1 0 0 1 1 1
+  Both x1 and x2 are larger than lo, but x1 > x2 since its first different
+  bit from lo is the 3rd one, while with x2 it is the 7th one. As a result,
+  if both x1 and x2 satisfy bits, x2 would be closer to our true result.
+
+  2. Formality:
+
+  Call r the smallest value not smaller than lo that satisfies bits. Since lo
+  does not satisfy bits, lo < r (2.1)
+
+  Call i the largest bit index such that:
+
+  - lo[x] satisfies bits for 0 <= x < i (2.2)
+  - zeros[i] = 0                        (2.3)
+  - lo[i]    = 0                        (2.4)
+
+  Consider v:
+
+  - v[x] = lo[x], for 0 <= x < i        (2.5)
+  - v[i] = 1                            (2.6)
+  - v[x] = ones[x], for x > i           (2.7)
+
+  We will prove that v == r.
+
+  a. Firstly, we prove that r <= v:
+
+    a.1. We have lo < v since:
+      lo[x] == v[x], for 0 <= x < i (according to 2.5)
+      lo[i] <  v[i] (according to 2.4 and 2.6, lo[i] == 0 < v[i] == 1)
+      bits at x > i have lower significance, and are thus irrelevant
+
+    a.2. We have v satisfies bits, this is because:
+      v[x] satisfies bits for 0 <= x < i (according to 2.2 and 2.5)
+      v[i] satisfies bits:
+        According to 2.3 and 2.6, v[i] == 1 and zeros[i] == 0, v[i] does not violate
```

Comment:
```suggestion
        According to 2.3 and 2.6, zeros[i] == 0 and v[i] == 1, v[i] does not violate
```
Keep the same order as the references.